2단계 - 미션 1차 구현

Eundms opened the pull request.

Eundms committed:
2단계 - 미션 1차 구현

javajigi commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Arrays;
```
패키지 없음
calculator와 같은 패키지 추가 후 구현하면 어떨까?

javajigi commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+
+public class StringAddCalculator {
+
+  private static String delimiter = ":|,";
```
```suggestion
  private static final String DELIMITER = ":|,";
```
상수 컨벤션 지키기

javajigi commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+      return 0;
+    }
+
+    Matcher m = Pattern.compile("//(.)\n(.*)").matcher(input);
```
`"//(.)\n(.*)"`도 상수로 구현하면 어떨까?

javajigi commented on src/main/java/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+      if(number < 0) throw new RuntimeException("음수");
+      result += number;
+    }
+    return result;
```
메서드가 너무 많은 일을 하고 있다.
한 가지 책임만 가지도록 리팩터링한다.

이 단계에서는 **단일 책임 원칙(SRP)**을 달성하기 위한 가장 강력한 무기, 바로 '메서드 분리(Extract Method)' 연습에 집중할 때입니다.

🎯 목표: 극단적인 메서드 분리를 통한 '설계 임계점' 돌파

이 단계의 가장 중요한 목표는 메서드가 오직 한 가지 책임만 가지도록 리팩터링하는 연습입니다.

현실 프로젝트에서는 유연성을 발휘할 수 있지만, 지금 이 교육 과정에서만큼은 이 원칙을 가장 극단적인 수준으로 적용해 보길 권장합니다.

예를 들어, **'단 한 단계의 들여쓰기만 허용'**이라는 제약이 대표적입니다. 이 제약은 단순히 코드를 예쁘게 만드는 규칙이 아닙니다. 이것은 여러분의 메서드에 **"너는 이 조건문 이상의 책임은 질 수 없다"**고 선언하는 강력한 설계 압박 수단입니다.

💡 극단적인 연습이 주는 두 가지 핵심 성장 동력

이처럼 '메서드 분리의 극한'에 도전하는 것이 단순한 고행이 아닌, **의식적인 연습(Deliberate Practice)**의 핵심인 이유가 있습니다.

1. 테스트 가능한 설계의 완성

메서드가 한 가지 책임만 가지게 될 때, 그 메서드는 단위 테스트(Unit Test) 하기에 완벽한 상태가 됩니다. 책임이 하나인 메서드는 입력과 출력이 명확해지고, 테스트를 위해 복잡한 환경 설정(Fixture)이 필요 없어지기 때문입니다. 테스트하기 어려운 코드는 곧 잘못 설계된 코드라는 깨달음을 얻게 될 것입니다.

2. '설계 임계점' 돌파와 새로운 패턴의 발견

익숙한 방식대로 두세 가지 역할을 한 메서드에 몰아넣으면 당장은 편합니다. 하지만 극한의 제약을 적용하는 순간, 기존의 코드는 무너지기 시작합니다. 그리고 이 무너진 코드를 다시 일으켜 세우기 위해 노력하는 과정에서 여러분은 필연적으로 새로운 설계 아이디어를 떠올리게 됩니다.

이것은 기존의 지식으로 해결할 수 없는 지점을 넘어서는 **'설계 임계점(Design Threshold)'**을 돌파하는 경험이며, 여러분이 이 과정을 통해 얻는 깨달음은 단순한 코드 조각 이상의 강력한 설계 근육이 될 것입니다.

🧑‍💻 지금, 도전하세요!

메서드 리팩터링의 '적정 수준'이 어느 정도일지는 지금 당장 정답을 찾기 힘듭니다. 따라서 이번에는 **"너무 과한데?"**라고 느껴질 정도로 분리해 보세요.

과하게 분리한 후, 왜 그 방법이 과했는지 스스로 고민하는 것이 가장 중요합니다. 이 과정 자체가 **'책임의 경계'**를 명확하게 판단하는 능력을 길러줄 것입니다.

javajigi commented on src/test/java/SetTest.java:
```diff
@@ -20,12 +21,17 @@ public void setUp() {
+
+  @Test
+  void check_set_size(){
+    assertThat(numbers.size()).isEqualTo(3);
```
```suggestion
        assertThat(numbers).hasSize(3);
```

collection의 경우 위와 같이 구현도 가능함

javajigi commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -3,37 +3,57 @@
+    int result = StringAddCalculator.splitAndSum("1");
+    assertThat(result).isEqualTo(1);
+  }
+
```
이 코드 아래 부분의 test case는 StringTest로 분리해야 하지 않을까?

javajigi reviewed (changes requested):
어제 1단계 리뷰 요청 후 merge하고 수업 시간이 끝난 후에 미션 진행하느라 넘 수고하셨어요. 👍 
바쁜 일상 업무를 진행하면서 학습을 병행하려고 노력하는 은정님의 모습이 느껴져 한편으론 안쓰럽기도 하고, 기특한 마음이 들기도 하네요.

은정님. 코드를 보니 미션에 투자할 시간이 정말 많지 않음이 느껴지네요. ㅠㅠㅠ
그런데 지금과 같이 미션을 끝내는 것에만 집중하다보면 실력이 빠르게 늘지 않을 것 같아요.
모든 미션을 완료하지 않아도 괜찮으니 약간은 조급함을 버리고 미션 진행해 보면 어떨까요?
오히려 약간의 여유를 가지고 미션에 집중해보면 즐거움도 느낄 때도 있고, 그럼 미션을 더 많은 투자하고 싶은 마음도 생길 수 있습니다.

이번 미션은 메서드 분리에 집중 투자하는 시간입니다.
splitAndSum 메서드 리팩터링하는데 시간 투자해 보면 어떨까요?
응원합니다.

Eundms committed:
2단계 - 문자열 덧셈 계산기

javajigi commented on src/main/java/calculator/Delimiter.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class Delimiter {
```
👍 
메서드 분리 연습에 집중하는 연습이라 이 정도까지 기대하지 않는데요.
와 이렇게 접근하다니 인상적이네요.

javajigi commented on src/main/java/calculator/Delimiter.java:
```diff
@@ -0,0 +1,31 @@
+  private final String delimiter;
+  private final String content;
+
+  private Delimiter(String delimiter, String content){
```
클래스 이름 한번 고민해 보면 어떨까요?
AI 추천해 준 이름 한번 제안해 봅니다.

클래스의 기능과 역할에 비추어 볼 때, InputParser 또는 InputContext가 가장 적합합니다.
- 만약 객체 생성(예: parse 메서드) 시 복잡한 분석 로직을 강조하고 싶다면: InputParser
- 만약 객체 생성 후 분리된 데이터(delimiter와 content)를 보관하고 사용하는 역할에 초점을 맞추고 싶다면: InputContext

javajigi commented on src/main/java/calculator/Numbers.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Arrays;
+import java.util.List;
+
+class Numbers {
```
일급 콜렉션 벌써 적용했군요. 👍

javajigi commented on src/main/java/calculator/Numbers.java:
```diff
@@ -0,0 +1,32 @@
+    this.values = values;
+  }
+
+  static Numbers from(String[] tokens){
```
굳이 정적 팩터리 메서드로 구현할 필요가 있을까?
이번 기회에 생성자와 정적 팩터리 메서드의 적절한 사용에 대해 고민해 보는 기회를 만들면 어떨까?

AI에게 정적 팩터리 메서드 단점에 대해 질문한 결과 공유해 봅니다.

🔍 발견성 (Discoverability) 문제
생성자와의 차이점: 생성자는 자바 API 문서(Javadoc)에서 클래스 이름 아래에 명확하게 표시됩니다. 개발자는 new 키워드를 사용하는 순간 IDE의 자동 완성 기능을 통해 사용 가능한 모든 생성자를 쉽게 확인할 수 있습니다.
정적 팩터리 메서드의 문제: 정적 팩터리 메서드는 일반적인 static 메서드와 동일하게 취급되므로, 개발자가 API 문서나 코드를 직접 살펴보지 않으면 해당 클래스가 객체 생성을 위한 특별한 메서드(팩터리 메서드)를 제공하는지 쉽게 알 수 없습니다.

완화 방법: 이펙티브 자바(Effective Java)에서 제시하는 from, of, valueOf, getInstance 등의 표준 명명 관례를 따르면 이 문제가 완화될 수 있지만, 여전히 개발자가 이 관례를 알고 있어야 합니다.

🚫 상속 제한
필요성: 클래스를 상속(확장)하려면 하위 클래스의 생성자가 상위 클래스의 public 또는 protected 생성자에 접근할 수 있어야 합니다.
정적 팩터리의 영향: 정적 팩터리 메서드를 사용하는 클래스는 객체 생성 로직을 캡슐화하고 재사용 가능한 인스턴스를 관리하기 위해 생성자를 private으로 선언하는 경우가 일반적입니다.

결과: 생성자가 private이면 외부에서 상속이 불가능해집니다. 즉, 정적 팩터리 메서드만 제공하는 클래스는 상속을 통한 확장이 불가능합니다.

참고: 이는 불변(Immutable) 객체를 설계하거나 상속보다 조합(Composition)을 권장하는 설계 원칙에서는 오히려 장점으로 간주되기도 합니다.

🧩 인터페이스 타입으로 구현할 수 없음
자바의 인터페이스는 정적 메서드를 가질 수 있지만 (Java 8 이후), 인터페이스의 인스턴스를 생성하는 정적 팩터리 메서드를 정의하는 것은 불가능합니다.
예외: 자바는 관례적으로 Collections.singletonList()처럼 **별도의 동반 클래스(companion class)**를 만들어 인터페이스의 인스턴스를 반환하는 정적 팩터리 메서드를 제공합니다. (예: java.util.Collections 클래스는 java.util.Collection 인터페이스를 위한 정적 팩터리 메서드들을 모아둔 클래스입니다.)

javajigi commented on src/main/java/calculator/Numbers.java:
```diff
@@ -0,0 +1,32 @@
+    }
+  }
+
+  int sum() {
```
👍

javajigi reviewed (changes requested):
이번 미션이 메서드 분리 연습에 집중하는 미션인데요.
벌써 클래스 분리까지 넘 인상적이네요. 💯 
특히 Delimiter 클래스 분리하기 위한 접근 👍 
바로 merge하려다 이번 기회에 생성자 vs 정적 팩터리 메서드 사용에 대해 고민해 봤으면 하는 바람에 한번 더 request chages로 피드백 합니다.
많은 개발자가 정적 팩터리 메서드 사용을 남용하는 경우가 있는데요.
정적 팩터리 메서드 사용보다 생성자를 사용하는 것으로 접근해 보면 어떨까요?

Eundms committed:
클래스별 테스트 분리

Eundms committed:
클래스 이름 변경

Eundms committed:
정적팩터리메서드 대신 생성자 사용

- 위의 경우 생성자 내부에서 처리하는 것이 좋음
- 정적 팩터리 메서드를 사용하는 것이 좋은 경우는 언제일까?
1) 같은 파라미터 타입으로 다른 의미의 객체 생성
2) 하위 타입을 반환하는 경우

javajigi reviewed (approved):
피드백 반영 잘 했네요. 👍 
앞으로도 생성자와 정적 팩터리 메서드 사용할 때 어느 방식으로 객체를 생성할 것인지 의식적으로 고민해 보면 좋겠습니다.

드디어 본격적인 자동차 경주 게임 구현이네요.
현재 은정님의 역량에 따라 어느 부분을 도전할 것인지 고민해 보면 좋겠습니다.
TDD가 생각보다 쉽지 않은 도전이라서요.
기존에 객체 설계, 단위 테스트를 고민해 본 경험이 많지 않다면 3단계는 객체 설계에 집중해도 괜찮아요.
어떤 방식으로 도전했는지 3단계 pr 보낼 때 공유해 주셔도 좋겠습니다.
